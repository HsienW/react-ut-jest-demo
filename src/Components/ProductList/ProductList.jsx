import React, {Component} from 'react';
import PropTypes from 'prop-types';

export class ProductList extends Component {

    state = {
        searchKey: ''
    };

    changeKey = (event) => {
        this.setState({
            searchKey: event.target.value
        });
    };

    search = () => {
        event.preventDefault();
        this.props.doListSearch(this.state.searchKey);
    };

    render() {
        const {listData} = this.props;
        return (
            <div>
                <h2>Product-list</h2>
                <input type="text" value={this.state.searchKey} onChange={this.changeKey}/>
                <button onClick={this.search}>Search</button>
                <ul>
                    {
                        listData.length !== 0
                            ? listData.map((item) => {
                                return (
                                    <li key={item.id}>{item.name}</li>
                                );
                            })
                            : <div>No data</div>
                    }
                </ul>
            </div>
        );
    }
}

ProductList.defaultProps = {
    listData: [],
    doListSearch: () => {}
};

ProductList.propTypes = {
    listData: PropTypes.array.isRequired,
    doListSearch: PropTypes.func.isRequired
};
