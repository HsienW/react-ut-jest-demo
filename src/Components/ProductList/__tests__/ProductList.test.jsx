import React from 'react';
import Enzyme, {shallow, mount} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import {ProductList} from '../ProductList';


Enzyme.configure({adapter: new Adapter()});


describe('ProductList component', () => {
    const wrapper = shallow(<ProductList listData={[]} doListSearch={() => {}}/>);

    it('should created component', () => {
        expect(wrapper.exists()).toBe(true);
    });

    it('should created snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    });
});

describe('ProductList props', () => {
    it('should get default props', () => {
        const props = {
            listData: [],
            doListSearch: () => {}
        };
        const wrapper = mount(<ProductList {...props}/>);
        expect(wrapper.prop('listData')).toEqual(props.listData);
        expect(wrapper.prop('doListSearch')).toEqual(props.doListSearch);
    });

    it('should get new props', () => {
        const props = {
            listData: [
                {
                    id: 1,
                    name: 'Apple'
                },
            ],
            doListSearch: () => {}
        };
        const wrapper = mount(<ProductList {...props}/>);
        expect(wrapper.find('ul').children()).toHaveLength(1);
    });
});

describe('ProductList function - search', () => {
    it('should change search key value', () => {
        const wrapper = mount(<ProductList listData={[]} doListSearch={() => {}}/>);
        wrapper.find('input').simulate('change', {target: {value: 'Apple'}});
        expect(wrapper.state('searchKey')).toEqual('Apple');
    });
});
