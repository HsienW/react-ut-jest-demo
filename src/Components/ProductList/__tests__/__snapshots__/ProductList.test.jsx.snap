// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductList component should created snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ProductList
    doListSearch={[Function]}
    listData={Array []}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h2>
          Product-list
        </h2>,
        <input
          onChange={[Function]}
          type="text"
          value=""
        />,
        <button
          onClick={[Function]}
        >
          Search
        </button>,
        <ul>
          <div>
            No data
          </div>
        </ul>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Product-list",
        },
        "ref": null,
        "rendered": "Product-list",
        "type": "h2",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "onChange": [Function],
          "type": "text",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Search",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "Search",
        "type": "button",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <div>
            No data
          </div>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "No data",
          },
          "ref": null,
          "rendered": "No data",
          "type": "div",
        },
        "type": "ul",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h2>
            Product-list
          </h2>,
          <input
            onChange={[Function]}
            type="text"
            value=""
          />,
          <button
            onClick={[Function]}
          >
            Search
          </button>,
          <ul>
            <div>
              No data
            </div>
          </ul>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Product-list",
          },
          "ref": null,
          "rendered": "Product-list",
          "type": "h2",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "onChange": [Function],
            "type": "text",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Search",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Search",
          "type": "button",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <div>
              No data
            </div>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "No data",
            },
            "ref": null,
            "rendered": "No data",
            "type": "div",
          },
          "type": "ul",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": false,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": true,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;
